### PR TITLE
Add mutable reference syntax to parser

### PR DESCRIPTION
**Pull request by DaveDuck321**
I'm working on full const-correctness atm, we'll need the parser changes merged separately to avoid breaking the bootstrap.

At the moment, I'm implementing the following rules:

1)
```rust
let a : int = 9;  // `a` is a constant
(&a); // Gives a constant reference to `a`
(&mut a); // Illegal: cannot take a mutable reference to constant `a`
```

2)
```rust
mut thing : int = 0;
let a : int mut& = &mut thing; // Syntax for mutable reference
mut a : int mut& = &mut thing; // Illegal: cannot store a reference in a mutable variable.
```

3)
```rust
typedef MyStruct : {a : int mut&}
let myStruct : MyStruct = ...
(&mut myStruct.a); // Legal??? Since we could always copy myStruct into mutable memory and assign to the same reference? Maybe we should change this tho?
```

4)
```rust
&mut T /* is implicitly convertible to */ &T // (but doesn't pattern match?)
```

5)
```rust
int[&, 9] // A constant reference heap array;
int[mut&, 9] // A mutable reference heap array
```

@antroseco Would like ur feedback regarding 3 tho, maybe we should change it just so its closer to c++. Also do you like the `mut`/ `let` syntax or should we go back to `let`/ `const`

### CODE REVIEW

**Comment by antroseco** [comment]
> I'm working on full const-correctness atm, we'll need the parser changes merged separately to avoid breaking the bootstrap.
> 
> At the moment, I'm implementing the following rules:
> 
> 
> ```rust
> let a : int = 9;  // `a` is a constant
> (&a); // Gives a constant reference to `a`
> (&mut a); // Illegal: cannot take a mutable reference to constant `a`
> ```

Reasonable

> ```rust
> mut thing : int = 0;
> let a : int mut& = &mut thing; // Syntax for mutable reference
> mut a : int mut& = &mut thing; // Illegal: cannot store a reference in a mutable variable.
> ```

Fair, so the references themselves are always const, which is consistent with our syntax (or lack thereof, as we have no way of modifying them)
 
> ```rust
> typedef MyStruct : {a : int mut&}
> let myStruct : MyStruct = ...
> (&mut myStruct.a); // Legal??? Since we could always copy myStruct into mutable memory and assign to the same reference? Maybe we should change this tho?
> ```

I'm reading this as "const struct with a reference to a mutable int". I don't see why it would be illegal to mutate the mutable int, as its the reference itself that should be const. This is similar to allowing `list.append()` to modify a frozen list in Python, or just `int *`/`int &` inside a `const` struct in C++.

> ```rust
> &mut T /* is implicitly convertible to */ &T // (but doesn't pattern match?)
> ```

Should this be an error instead? You should be able to re-borrow your `&mut` as a `&` to make it go away of course, but on the other hand, this might make generic code harder, as all functions you call with your generic type need to consistently take `&` or `&mut`.
 
> ```rust
> int[&, 9] // A constant reference heap array;
> int[mut&, 9] // A mutable reference heap array
> ```

Would we be able to reassign a `mut array : int[&]` then? 
If yes, `mut array : int[&mut]` is also valid, right?

> @antroseco Would like ur feedback regarding 3 tho, maybe we should change it just so its closer to c++.

I think this is already C++ behaviour, no?

> Also do you like the `mut`/ `let` syntax or should we go back to `let`/ `const`

Yeah, I've been thinking about this myself, and I've been leaning more towards `mut`/`let` too

File: src/glang/parser/lexer_parser.py [hunk 1]
```diff
@@ -66,7 +66,7 @@ class ArrayType(Type):
 
 @dataclass
 class HeapArrayType(ArrayType):
-    pass
+    is_const: bool
```
[the remark this answers is not in the document]
Comment (DaveDuck321) [reply src/glang/parser/lexer_parser.py]: Hmmm, maybe it matches our grammar better.

Done in [d404260](https://github.com/DaveDuck321/GrapheneLang/pull/119/commits/d404260202c6a10eaa6cbcb6e73f36a936238a52).

File: src/glang/parser/parser.c3 [hunk 2]
```diff
@@ -1086,9 +1103,9 @@ function parse_const_declaration : (cursor : Cursor&) -> VariableDeclaration = {
     return {true, variable_name, type, make_optional(expression), meta(start_offset, cursor:skip_ws_return_offset())};
 }
 
-function parse_let_declaration : (cursor : Cursor&) -> VariableDeclaration = {
+function parse_mut_declaration : (cursor : Cursor&) -> VariableDeclaration = {
```
[the remark this answers is not in the document]
Comment (DaveDuck321) [reply src/glang/parser/parser.c3:1109]: Yeah, I'll do this with with the main commit

File: src/glang/codegen/builtin_types.py [hunk 3]
```diff
@@ -89,11 +89,10 @@ def get_name(self) -> str:
         return f"{self.name}{specialization}"
 
     def format_for_output_to_user(self, full=False) -> str:
-        reference_suffix = (
-            "&"
-            if self.is_reference and not self.definition.is_always_a_reference
-            else ""
-        )
+        if self.definition.is_always_a_reference:
+            reference_suffix = "" if self.is_reference else "unborrowed"
```
Comment (antroseco) [inline src/glang/codegen/builtin_types.py:93]: Shouldn't this also be `" (unborrowed)"` or similar?

**Comment by antroseco** [comment]
> > > ```rust
> > > int[&, 9] // A constant reference heap array;
> > > int[mut&, 9] // A mutable reference heap array
> > > ```
> > 
> > 
> > Would we be able to reassign a `mut array : int[&]` then? If yes, `mut array : int[&mut]` is also valid, right?
> 
> We can't reassign a heap array since it is just a reference to an array. The syntax doesn't really make a lot of sense either:
> 
> ```
> function main : () -> int = {
>     let a : int[3] = {7, 9, 7};
>     let b : int[4] = {5, 3, 2, 1};
> 
>     let ref_a : int[&] = &a;
>     let ref_b : int[&] = &b;
> 
>     ref_a = &ref_b;
>     return 0;
> }
> ```
> 
> ^ Gives a nonsense type error: `int[&] != int[&]` which I fixed: `Error in assignment: type 'int[&]' does not match expected type 'int[&] (unborrowed)'`

Uhh I thought the syntax made sense, but on the other hand, you're right it would be inconsistent with normal references.

> ```
> function main : () -> int = {
>     let a : int[3] = {7, 9, 7};
>     let b : int[4] = {5, 3, 2, 1};
> 
>     let ref_a : int[&] = &a;
>     let ref_b : int[&] = &b;
> 
>     ref_a = ref_b;
>     return 0;
> }
> ```
> 
> ^ Triggers an assert
> 
> We should work on making heap arrays behave more predictably tbh

Yep

Regarding bootstrapping, I'm not sure why you chose to disable some of the checks in the parser. Shouldn't we just add a new stage to `bootstrap.sh`?
- Bootstrap up to the previous tag
- Build a parser that uses the old syntax but can parse the new syntax (new tag)
- Build a parser that uses the new syntax and can parse the new syntax (working tree)

**Comment by DaveDuck321** [comment]
> Regarding bootstrapping, I'm not sure why you chose to disable some of the checks in the parser. Shouldn't we just add a new stage to `bootstrap.sh`?
> 
>     * Bootstrap up to the previous tag
> 
>     * Build a parser that uses the old syntax but can parse the new syntax (new tag)
> 
>     * Build a parser that uses the new syntax and can parse the new syntax (working tree)

Are you saying that you want us to keep this commit in some arbitrary branch with the bootstrap tag and not bother merging it into main?

Alternatively, we could merge this commit into main as is, we'd lose `const`  for a single commit before `let` starts implying the same thing.

File: src/glang/codegen/builtin_types.py (hunk 3 above)
Comment (DaveDuck321) [reply src/glang/codegen/builtin_types.py:93]: I've fixed this in #126. We now print `u8[<unborrowed>]` in the same situation.

**Comment by antroseco** [comment]
> Alternatively, we could merge this commit into main as is, we'd lose `const` for a single commit before `let` starts implying the same thing.

Yeah this is fine

**Pull request merged**